Convert the SSE server to a Twisted pika connection

jeremycline opened the pull request:
Prior to this commit, a blocking connection to RabbitMQ was used. This
is problematic in Twisted because it blocks the reactor rather than take
advantage of Twisted's reactor design.

This should behave almost exactly like the previous connection, except
that now messages are pushed by RabbitMQ to the SSE server and in turn
immediately pushed out to clients. It also works with a single TCP
connection rather than using one connection per queue.

Fixes #48 and #50

Signed-off-by: Jeremy Cline jeremy@jcline.org

A few notes for reviewers:
- I'm really sorry for the huge diff. I didn't want it to be so large, but I couldn't just drop the new non-blocking connection without touching all the things that used it.
- I've based this branch off a branch that hasn't been merged quite yet so right now it's pulling in a few additional commits. The last one is the only one of interest here.
- This PR is currently a work in progress, but I wanted to give people an opportunity to look at it and comment early on. This should be functional, if a little unpolished. There are quite a few `# TODO` comments, and they'll all go away. I'll make a note when I think it's ready for a final review.
- I tried to translate 1-to-1 what was there previously with what's here now, so it should behave the same way in terms of exchanges, queues, and their configurations.

Let me know what you all think!

skrzepto committed:
setting status code to 404 when the route is not found

jeremycline committed:
Add a 404 error page class that handles JSON bodies

By default, this class renders as HTML, but it accepts a ``detail``
parameter which is JSON-serialized and used as the response body.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Convert the SSE server to a Twisted pika connection

Prior to this commit, a blocking connection to RabbitMQ was used. This
is problematic in Twisted because it blocks the reactor rather than take
advantage of Twisted's reactor design.

This should behave almost exactly like the previous connection, except
that now messages are pushed by RabbitMQ to the SSE server and in turn
immediately pushed out to clients. It also works with a single TCP
connection rather than using one connection per queue.

Fixes #48 and #50

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Rename and refactor various methods and classes

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline commented:
I've updated this some and added unit tests.

There are still a lot of TODOs, and I think if I address them all this PR will get very large and even more complex. After people have reviewed and approved this, I will file each TODO left as an issue and fix them individually. They range from small, nice-to-have features to problems with DoS to proper error handling.

skrzepto commented on fmn/sse/server.py:
```diff
@@ -0,0 +1,301 @@
+            auto_delete=False,
+            arguments={'x-message-ttl': self.expire_ms}
+        )
+        # TODO investigate why this routing_key is used
```
to prevent collision between a user and a group name having the same identifier.

jeremycline replied:
That makes sense, but does this work? The queue name is still the same, so currently what happens is if you request `/user/jcline/` and `/group/jcline/` you get a queue, `jcline` with two routing keys: `user-jcline` and `group-jcline`.

I'm still pretty new working with AMQP, but doesn't that mean any message with either of those routing keys will be routed to the queue? I'd get both the group messages and user messages in one queue, right?

skrzepto replied:
Im new to it too but keep in mind the exchange is also considered here as well. so its 

exchange: user, queue: jcline
exchange: group, queue: jcline

or at least that was my goal. So the routing key might not be necessary (might need to look back at rabbitmq docs for the reason why the routing key was used)

jeremycline replied:
So I gave this a shot and it seems that what I got is one queue, `jcline`, that's bound to the exchange `user` with a routing key of `user-jcline` and the exchange `group` with a routing key of `group-jcline`. Messages published to either exchange with the appropriate routing key end up in `jcline`.

I _think_ we can avoid collision between users and groups by having a queue for each. That is, a `group-jcline` queue (presumably for my fan group) and the `user-jcline` queue for me. Right now I can't think of a good use for having a routing key. Using the queue name _should_ work (I'm waving my hands wildly right now, I haven't tried this).

skrzepto replied:
https://www.rabbitmq.com/tutorials/tutorial-four-python.html

It seems for Direct Exchange 

```
The routing algorithm behind a direct exchange is simple - 
a message goes to the queues whose binding key exactly 
matches the routing key of the message.
```

But i think we are using subscribe here instead which the example below uses a routing key so im a bit confused by this as well

jeremycline replied:
I believe the current implementation doesn't work as intended, but I propose that it is not fixed in this PR since this simply carries over previous behaviour. I'll write a ticket for this particular problem.

skrzepto replied:
Do you think the current setup for rabbitmq subscribe supports fanout? (fanout is if node A and B are listening to the same queue they both get the same messages) because if we do theoretically hit the ceiling on concurrent active connections for a single program we can always scale horizontally by spawning more the `server.py` processes and having a load balancer direct traffic if needed. Just something to think about for further improvements

jeremycline replied:
I believe `fanout` ignores the routing key and routes any message to the exchange to all queues[0]. This means that users of two _different_ queues bound to a fanout exchange both get the same messages.

I can think of a few approaches to scale horizontally. I seem to recall that if haproxy has access to the HTTP request (that is, it's proxying HTTP rather than HTTPS, or if it acts as the TLS endpoint) you can set it to route requests for the same resource to the same backend server, but that backend server is initially chosen at random. It's been a long time since I read the haproxy docs, though.

[0] https://www.rabbitmq.com/tutorials/amqp-concepts.html

skrzepto commented:
I know it was mentioned this was work in progress but was it able to send data? when I tried to `curl http://localhost:8080/user/skrzepto` with the master branch it displays messages like expected 

```
data: {"date_time": "2016-10-20T22:19:23.624103+00:00", "markup": "New message added to the queue.\ttime is: 18:19:23.624009  here's a random number: 0", "secondary_icon": "https://secure.gravatar.com/avatar/ba1882fd5522d16213b0535934f77b796f0b89f76edd65460078099fe97c20ea.jpg?s=64&d", "dom_id": "80d84850-ba46-413b-8de2-00a3635d104d", "link": "localhost:8080", "icon": "https://seccdn.libravatar.org/avatar/ba1882fd5522d16213b0535934f77b796f0b89f76edd65460078099fe97c20ea?s=64&d=retro"}
```

but once i shut down that instance and run this branch nothing is going through. I ran the program like this `python fmn/sse/server.py` Is there something im missing?

jeremycline commented:
Are there messages in the queue? Does it at least send the HTTP headers?

skrzepto commented on fmn/sse/server.py:
```diff
@@ -0,0 +1,301 @@
+            subscription = self.subscribers[request_key]
+            subscription['queue'] = rabbit_queue
+
+            while True:
```
It seems like it never reaches here which is why messages aren't being sent

skrzepto commented on fmn/sse/server.py:
```diff
@@ -0,0 +1,301 @@
+            subscription['queue'] = rabbit_queue
+
+            while True:
+                channel, method, properties, body = yield rabbit_queue.queue.get()
```
It seems that this never really returns and doesnt reach the code which is why messages aren't being sent

jeremycline replied:
Can you fpaste the log output?

jeremycline replied:
Actually I think I know what the problem is. I removed the bit where it appended '.id.fedoraproject.org' if it was missing and went with requiring the full queue name. Does `http://localhost:8080/user/skrzepto.id.fedoraproject.org` work?

skrzepto replied:
works now :) thanks

skrzepto commented:
Overall this PR looks good I think we should close the other two and merge this one. Let me check performance benchmark but Im assuming its going to be way faster then it was previously :)

jeremycline commented:
Out of interest, what tool are you using to benchmark this?

skrzepto commented:
I use `wrk` but heard good things about `ab` (apache benchmark). Keep in mind of your ulimit on your system though. My experience is if you request 10,000 connections but your systems ulimit is 1024 only 1024 connections go through the bench mark but it continues without fail. This is why I was printing out the length of the requests to make sure they were all coming through

skrzepto commented:
current benchmark is 4000 connections to one route its sitting at 100% cpu utilization and 20% ram on the lowest end digital ocean server

Looks like the cpu usage is more stabalized then before. I'll try to push it to 10,000 another day

```
./wrk -t2 -c4000 -d300s http://localhost:8080/user/skrzepto.id.fedoraproject.org                               [20:16:29]
Running 5m test @ http://localhost:8080/user/skrzepto.id.fedoraproject.org
  2 threads and 4000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     0.00us    0.00us   0.00us    -nan%
    Req/Sec     0.00      0.00     0.00      -nan%
  0 requests in 5.00m, 2.40GB read
  Socket errors: connect 0, read 122, write 115, timeout 0
Requests/sec:      0.00
Transfer/sec:      8.19MB


```

skrzepto commented:
After the long stress test this error popped up but didnt crash it

```
Unhandled error in Deferred:


Traceback (most recent call last):
  File "/root/repo/fmn.sse/sse/lib/python2.7/site-packages/twisted/web/http.py", line 1951, in connectionLost
    request.connectionLost(reason)
  File "/root/repo/fmn.sse/sse/lib/python2.7/site-packages/twisted/web/http.py", line 1337, in connectionLost
    d.errback(reason)
  File "/root/repo/fmn.sse/sse/lib/python2.7/site-packages/twisted/internet/defer.py", line 434, in errback
    self._startRunCallbacks(fail)
  File "/root/repo/fmn.sse/sse/lib/python2.7/site-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/root/repo/fmn.sse/sse/lib/python2.7/site-packages/twisted/internet/defer.py", line 587, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "fmn/sse/server.py", line 243, in request_closed
    subscriber['queue'].queue.close()
exceptions.TypeError: close() takes exactly 2 arguments (1 given)
```

jeremycline committed:
Close the ClosableDeferredQueue with an error message

pika provides a method to close the DeferredQueue which pops all waiting
requests to push or get and calls errback on them with the provided
reason.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline commented:
If you don't have objections, I'd like to turn the TODOs into tickets and merge this PR.

skrzepto commented:
@jeremycline fine with me but we should merge this pr in first before making the tickets
